src/datamodel/builder/UMessage.cpp

```diff
@@ -10,3 +10,224 @@
+        "Request id is not a valid UUid | " +
+        std::string(UUidValidator::message(*reqIdReason)));
+    }
+
+    UMessageBuilder builder(v1::UMessageType::UMESSAGE_TYPE_RESPONSE,
+                            std::move(method), std::move(sink),
+                            std::move(request_id));
+
+    return builder.withPriority(priority);
+}
+
+UMessageBuilder UMessageBuilder::response(const v1::UMessage& request) {
+    v1::UUri sink = request.attributes().source();
+    v1::UUID reqId = request.attributes().reqid();
+    v1::UPriority priority = request.attributes().priority();
+    v1::UUri method = request.attributes().sink();
+
+    return UMessageBuilder::response(std::move(sink), std::move(reqId),
+                                      priority, std::move(method));
+
+}
+
+UMessageBuilder& UMessageBuilder::withPriority(v1::UPriority priority) {
+    if ((priority < v1::UPriority::UPRIORITY_CS0) ||
+        (priority > v1::UPriority::UPRIORITY_CS6)) {
+        throw std::out_of_range("Priority value is out of range");
+    }
+
+    if ((attributes_.type() == v1::UMessageType::UMESSAGE_TYPE_REQUEST) ||
+        (attributes_.type() == v1::UMessageType::UMESSAGE_TYPE_RESPONSE)) {
+        if (priority < v1::UPriority::UPRIORITY_CS4) {
+            throw std::out_of_range("Priority value is out of range, for req/resp should be > CS4");
+        }
+    }
+
+    attributes_.set_priority(priority);
+
+    return *this;
+}
+
+UMessageBuilder& UMessageBuilder::withTtl(std::chrono::milliseconds ttl) {
+    if ((ttl.count() <= 0) || (ttl.count() > std::numeric_limits<uint32_t>::max())) {
+        throw std::out_of_range("TTL value is out of range");
+    }
+
+    attributes_.set_ttl(static_cast<uint32_t>(ttl.count()));
+
+    return *this;
+}
+
+UMessageBuilder& UMessageBuilder::withToken(const std::string& token) {
+    if ((attributes_.type() != v1::UMessageType::UMESSAGE_TYPE_REQUEST)) {
+        throw std::domain_error("Token can only be set on a request message");
+    }
+
+    attributes_.set_token(token);
+
+    return *this;
+}
+
+UMessageBuilder& UMessageBuilder::withPermissionLevel(int32_t level) {
+    if ((attributes_.type() != v1::UMessageType::UMESSAGE_TYPE_REQUEST)) {
+        throw std::domain_error("Permission level can only be set on a request message");
+    }
+
+    attributes_.set_permission_level(level);
+
+    return *this;
+}
+
+UMessageBuilder& UMessageBuilder::withCommStatus(v1::UCode code) {
+    if (attributes_.type() != v1::UMessageType::UMESSAGE_TYPE_RESPONSE) {
+        throw std::domain_error("CommStatus can only be set on a response message");
+    }
+
+    if (code != v1::UCode::OK) {
+        attributes_.set_commstatus(code);
+    }
+
+    return *this;
+}
+
+UMessageBuilder& UMessageBuilder::withPayloadFormat(v1::UPayloadFormat format) {
+    if ((format < v1::UPayloadFormat::UPAYLOAD_FORMAT_UNSPECIFIED) ||
+        (format > v1::UPayloadFormat::UPAYLOAD_FORMAT_SHM)) {
```

Comment:
Same as before but for UPayload, I suggest UPayloadFormat_MIN & UPayloadFormat_MAX

Reply:
I agree.

Done